hail/src/main/scala/is/hail/backend/BackendRpc.scala

```diff
@@ -0,0 +1,251 @@
+    ) extends Command
+  }
+
+  trait Ask[Env] {
+    def command(env: Env): Command
+  }
+
+  trait Context[Env] {
+    def scoped[A](env: Env)(f: ExecuteContext => A): (A, Timings)
+  }
+
+  trait Write[Env] {
+    def timings(env: Env)(t: ExecutionTimer.Timings): Unit
+    def result(env: Env)(r: Array[Byte]): Unit
+    def error(env: Env)(t: Throwable): Unit
+  }
+
+  implicit val fmts: Formats = DefaultFormats
+
+  import Commands._
+
+  final def runRpc[A](env: A)(implicit Ask: Ask[A], Context: Context[A], Write: Write[A]): Unit =
+    try {
+      val command = Ask.command(env)
+      val (result, timings) = retryTransientErrors {
+        Context.scoped(env) { ctx =>
+          command match {
+            case TypeOf(kind, s) =>
+              jsonToBytes {
+                (kind match {
+                  case Value => IRParser.parse_value_ir(ctx, s)
+                  case Table => IRParser.parse_table_ir(ctx, s)
+                  case Matrix => IRParser.parse_matrix_ir(ctx, s)
+                  case BlockMatrix => IRParser.parse_blockmatrix_ir(ctx, s)
+                }).typ.toJSON
+              }
+
+            case Execute(s, fns, codec) =>
+              val bufferSpec = BufferSpec.parseOrDefault(codec)
+              withIRFunctionsReadFromInput(ctx, fns) {
+                val ir = IRParser.parse_value_ir(ctx, s)
+                val res = ctx.backend.execute(ctx, ir)
+                res match {
+                  case Left(_) => Array.empty[Byte]
+                  case Right((pt, off)) =>
+                    using(new ByteArrayOutputStream()) { os =>
+                      Backend.encodeToOutputStream(ctx, pt, off, bufferSpec, os)
+                      os.toByteArray
+                    }
+                }
+              }
+
+            case ParseVcfMetadata(path) =>
+              jsonToBytes {
+                val metadata = LoadVCF.parseHeaderMetadata(ctx.fs, Set.empty, TFloat64, path)
+                Extraction.decompose(metadata)
+              }
+
+            case ImportFam(path, isQuantPheno, delimiter, missing) =>
+              jsonToBytes {
+                LoadPlink.importFamJSON(ctx.fs, path, isQuantPheno, delimiter, missing)
+              }
+
+            case LoadReferencesFromDataset(path) =>
+              jsonToBytes {
+                val rgs = ReferenceGenome.fromHailDataset(ctx.fs, path)
+                ctx.References ++= rgs.map(rg => rg.name -> rg)
+                JArray(rgs.map(_.toJSON).toList)
+              }
+
+            case LoadReferencesFromFASTA(name, fasta, index, xContigs, yContigs, mtContigs, par) =>
+              jsonToBytes {
+                val rg = ReferenceGenome.fromFASTAFile(
+                  ctx,
+                  name,
+                  fasta,
+                  index,
+                  xContigs,
+                  yContigs,
+                  mtContigs,
+                  par,
+                )
+                ctx.References += rg.name -> rg
```

Comment:
Previously this command didn't add the reference to the backend state (See BackendServer and ServiceBackendAPI). It seemed to me that it should. Thoughts?